File: src/main/java/io/github/colintimbarndt/chat_emotes/data/unicode/joiner/MatchTransformer.java
package io.github.colintimbarndt.chat_emotes.data.unicode.joiner;

import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.MatchResult;
import org.jetbrains.annotations.NotNull;

public interface MatchTransformer {
    void transform(@NotNull MatchResult r, @NotNull StringBuilder result);
    default @NotNull String transform(@NotNull MatchResult r) {
        final var builder = new StringBuilder();
        transform(r, builder);
        return builder.toString();
    }
    default boolean acceptsWidth(int w) {
        return w == acceptedWidth();
    }
    int acceptedWidth();
    default boolean requiresName() {
        return false;
    }
}

File: src/main/kotlin/io/github/colintimbarndt/chat_emotes/data/unicode/joiner/MatchTransformer.kt
package io.github.colintimbarndt.chat_emotes.data.unicode.joiner

import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.MatchResult

interface MatchTransformer {
    fun transform(r: MatchResult, result: StringBuilder)

    fun acceptsWidth(w: Int): Boolean = w == acceptedWidth

    val acceptedWidth: Int

    val requiresName: Boolean
        get() = false
}